### PR TITLE
add multiple channel option to Orchestra unicast slotframe

### PR DESCRIPTION
**Pull request by atiselsts**
This increases the performance of Orchestra by reducing the number of collisions, especially helpful in dense networks.

### CODE REVIEW

**Comment by atiselsts** [comment]
Using multiple channel offsets for the EB rule also feels like a good idea. What do you think? I will send in out for testing it in one of my projects.

**Comment by atiselsts** [comment]
I agree the hash function should be improved. I'd prefer using `sax()` from MSF or another more complex hash function. I would also prefer to make it time variant, if we go this way- for example, take into account the ASN besides the MAC address.

New attribute for the channel could work, I suppose.
Do you think the logic in tsch-slot-operation should be like this (where `packet_attr_channel_offset` is the attribute value set by the scheduler):
```
if(current_packet == NULL || packet_attr_channel_offset == 0xffff) {
  channel_offset = current_link->channel_offset;
} else {
  channel_offset = packet_attr_channel_offset;
}
```

**Comment by simonduq** [comment]
Yes exactly, right before calling `tsch_calculate_channel` (and only when `burst_link_scheduled` is 0).

I have nothing against adding ASN to the hash function nor providing SAX as a readily-available option!

**Comment by atiselsts** [comment]
I've changed the code as requested, moving the selection functionality to the Orchestra module.

For now the hashing remains simple - I realized that as the channel offset selection code is moved to Orchestra, time-variant hash function is not easy to do anymore. I'll leave that for future improvements.

**Comment by g-oikonomou** [comment]
Changed base to branch `release-4.4`. Rebase possibly required.

File: os/net/mac/tsch/tsch-slot-operation.c [hunk 1]
```diff
@@ -238,10 +238,20 @@ tsch_release_lock(void)
 
 /* Return channel from ASN and channel offset */
 uint8_t
-tsch_calculate_channel(struct tsch_asn_t *asn, uint8_t channel_offset)
+tsch_calculate_channel(struct tsch_asn_t *asn, uint16_t channel_offset, struct tsch_packet *p)
```
Comment (simonduq) [inline os/net/mac/tsch/tsch-slot-operation.c]: could we keep `channel_offset` on 8 bits and keep the check that it's no greater than 15? The code below can still read the packetbuf attribute, and assign it to `channel_offset` only if != 0xffff

File: os/net/mac/tsch/tsch-schedule.c [hunk 2]
```diff
@@ -224,9 +224,6 @@ tsch_schedule_add_link(struct tsch_slotframe *slotframe,
     if(timeslot > (slotframe->size.val - 1)) {
       LOG_ERR("! add_link invalid timeslot: %u\n", timeslot);
       return NULL;
-    } else if(channel_offset > 15) {
-      LOG_ERR("! add_link invalid channel_offset: %u\n", channel_offset);
-      return NULL;
```
Comment (simonduq) [inline os/net/mac/tsch/tsch-schedule.c]: I would like it if we could keep this check

File: os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-ns.c [hunk 3]
```diff
@@ -81,6 +80,12 @@ select_packet(uint16_t *slotframe, uint16_t *timeslot)
     if(timeslot != NULL) {
       *timeslot = get_node_timeslot(dest);
     }
+#if ORCHESTRA_MULTIPLE_CHANNEL_OFFSETS
+    /* set per-packet channel offset */
+    if(channel_offset != NULL) {
+      *channel_offset = ORCHESTRA_LINKADDR_HASH(dest) % (ORCHESTRA_MAX_HASH + 1);
```
Comment (simonduq) [inline os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-ns.c]: likewise, this could be % 16

**Comment by simonduq** [comment]
One more thought: I think the choffset selection should not be between 0 and 15 but rather within a given limited range, assigned to the orchestra rule. This would preserve the property that orchestra slotframes are all isolated on their own (set of) channel offset(s).

File: os/net/mac/tsch/tsch-slot-operation.c (hunk 1 above)
Comment (atiselsts) [reply os/net/mac/tsch/tsch-slot-operation.c]: Channel offset is already 16 bits in the `struct tsch_link` which is passed as a parameter when this function is called (`tsch_calculate_channel(&tsch_current_asn, current_link->channel_offset, current_packet);`) so restricting it to 8 bits in the parameter seems like an oversight. I would like to keep the change to `uint16_t` (or if you're absolutely sure, change the value in `struct tsch_link` and other places to 8 bits)

File: os/net/mac/tsch/tsch-schedule.c (hunk 2 above)
Comment (atiselsts) [reply os/net/mac/tsch/tsch-schedule.c]: I wonder why? There can be hopping sequences with more than 16 elements. I don't think standard says anything about this. Plus, many of the current hopping sequences have less than 16 elements, including the default one! As you know, with hopping sequence 4 elements long, the channel offsets 1, 5, 9 and 13 are all going to map to the same physical channel - so why not allow channel offsets 17, 21 etc. as well?

**Comment by atiselsts** [comment]
> One more thought: I think the choffset selection should not be between 0 and 15 but rather within a given limited range, assigned to the orchestra rule. This would preserve the property that orchestra slotframes are all isolated on their own (set of) channel offset(s).

I like the idea of reserving "own" channel offsets for Orchestra slotframes. One simple way would be to use the range starting from `slotframe_handle` value for the unicast slotframe, in this case no extra config would be needed.

**Comment by simonduq** [comment]
> I like the idea of reserving "own" channel offsets for Orchestra slotframes. One simple way would be to use the range starting from slotframe_handle value for the unicast slotframe, in this case no extra config would be needed.

Yes I'm not sure what the best way to configure. One thing is the unicast slotframe doesn't come last as we it to take precedence over default_common

File: os/net/mac/tsch/tsch-schedule.c (hunk 2 above)
Comment (simonduq) [reply os/net/mac/tsch/tsch-schedule.c]: well yes you're right; I had been somehow assuming that the standard mandated this but doesn't seem to be any other constraint than it being 16bit unsigned :)

**Comment by simonduq** [comment]
Any update on the choffset per-rule allocation @atiselsts ? Any hope to have this soon enough for the release?

**Comment by atiselsts** [comment]
I'm not sure about the API yet. Do you have a good idea how the channel offset range should be configured / decided?

It would be nice to have this feature in 4.4. Testing showed quite an improvement compared with the single channel offset approach.

**Comment by simonduq** [comment]
Maybe the current thing that goes through `ORCHESTRA_RULES` and assigne offsets 0, 1, 2... needs to be replaced with something else. Maybe each Orchestra rule needs its own define? And for the ones with multiple channel offsets, both a MIN and MAX defs?

File: os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c [hunk 4]
```diff
@@ -74,6 +74,17 @@ get_node_timeslot(const linkaddr_t *addr)
   }
 }
 /*---------------------------------------------------------------------------*/
+static uint16_t
+get_node_channel_offset(const linkaddr_t *addr)
+{
+  if(addr != NULL && ORCHESTRA_UNICAST_MAX_CHANNEL_OFFSET > ORCHESTRA_UNICAST_MIN_CHANNEL_OFFSET) {
```
Comment (simonduq) [inline os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c]: shouldn't we allow `ORCHESTRA_UNICAST_MAX_CHANNEL_OFFSET == ORCHESTRA_UNICAST_MAX_CHANNEL_OFFSET` ?

Comment (atiselsts) [reply os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-storing.c]: True that. Updated now...

**Comment by simonduq** [comment]
Awesome, thanks!

**Pull request merged**